File: agni-core/src/main/kotlin/dev/agni/core/protocol/Response.kt
package dev.agni.core.protocol

sealed class Response {
    object Pong : Response()
    object Ok : Response()
    data class Value(val value: ByteArray) : Response()
    object Null : Response()
    data class Error(val message: String) : Response()

    fun toBytes(): ByteArray = when (this) {
        is Pong -> "PONG".toByteArray(Charsets.UTF_8)
        is Ok -> "OK".toByteArray(Charsets.UTF_8)
        is Value -> value
        is Null -> "NULL".toByteArray(Charsets.UTF_8)
        is Error -> "ERR $message".toByteArray(Charsets.UTF_8)
    }
}

File: protocol/response.go
package protocol

import "fmt"

type ResponseKind int

const (
	Pong ResponseKind = iota
	Ok
	Value
	Null
	Error
)

type Response struct {
	Kind    ResponseKind
	Value   []byte
	Message string
}

func (r Response) Bytes() []byte {
	switch r.Kind {
	case Pong:
		return []byte("PONG")
	case Ok:
		return []byte("OK")
	case Value:
		return r.Value
	case Null:
		return []byte("NULL")
	case Error:
		return []byte(fmt.Sprintf("ERR %s", r.Message))
	default:
		return nil
	}
}
